PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py

```diff
@@ -0,0 +1,200 @@
+    if distance_type == 'chessboard':
+        cost = [1, 1, 1, 1, 1, 1, 1, 1]
+    elif distance_type == 'eculidean':
+        cost = [1, np.sqrt(2), 1, np.sqrt(2), 1, np.sqrt(2), 1, np.sqrt(2)]
+    else:
+        print('Unsupported distance type.')
+        return
+
+    T = float('inf') * np.ones_like(gridmap, dtype=np.float)
+    T[src[0], src[1]] = 0
+    if transform_type == 'distance':
+        eT = np.zeros_like(gridmap)
+    elif transform_type == 'path':
+        eT = ndimage.distance_transform_cdt(1-gridmap, distance_type)
+    else:
+        print('Unsupported transform type.')
+        return
+
+    # set obstacle T value to infinity
+    for i in range(nrows):
+        for j in range(ncols):
+            if gridmap[i][j] == 1.0:
+                T[i][j] = float('inf')
+    is_visited = np.zeros_like(T, dtype=bool)
+    is_visited[src[0], src[1]] = True
+    queue = [src]
+    calculated = [(src[0]-1)*ncols + src[1]]
+
+    while queue != []:
+        i, j = queue.pop(0)
+        for k, inc in enumerate(order):
+            ni = i + inc[0]
+            nj = j + inc[1]
+            if ni >= 0 and ni < nrows and nj >= 0 and nj < ncols \
+                    and not gridmap[ni][nj]:
+                is_visited[i][j] = True
+                T[i][j] = min(T[i][j], T[ni][nj] + cost[k] + alpha*eT[ni][nj])
+                if not is_visited[ni][nj] and ((ni-1)*ncols + nj) \
+                        not in calculated:
+                    queue.append((ni, nj))
+                    calculated.append((ni-1)*ncols + nj)
+
+    return T
+
+
+def wavefront(T, start, goal):
+    """wavefront
+
+    performing wavefront coverage path planning
+
+    :param T: the transform matrix
+    :param start: start point of planning
+    :param goal: goal point of planning
+    """
+
+    path = []
+
+    nrows, ncols = T.shape
+
+    if start[0] >= goal[0] and start[1] >= goal[1]:
+        order = [[1, 0], [0, 1], [-1, 0], [0, -1],
+                 [1, 1], [1, -1], [-1, 1], [-1, -1]]
+    elif start[0] <= goal[0] and start[1] >= goal[1]:
+        order = [[-1, 0], [0, 1], [1, 0], [0, -1],
+                 [-1, 1], [-1, -1], [1, 1], [1, -1]]
+    elif start[0] >= goal[0] and start[1] <= goal[1]:
+        order = [[1, 0], [0, -1], [-1, 0], [0, 1],
+                 [1, -1], [-1, -1], [1, 1], [-1, 1]]
+    elif start[0] <= goal[0] and start[1] <= goal[1]:
+        order = [[-1, 0], [0, -1], [0, 1], [1, 0],
+                 [-1, -1], [-1, 1], [1, -1], [1, 1]]
+    else:
+        return
+    cur = start
+    is_visited = np.zeros_like(T, dtype=bool)
+
+    while cur != goal:
+        i, j = cur
+        path.append((i, j))
+        is_visited[i][j] = True
+
+        max_T = float('-inf')
+        imax = (-1, -1)
+        ilast = 0
+        for ilast in range(len(path)):
+            cur = path[-1-ilast]
+            for ci, cj in order:
+                ni, nj = cur[0] + ci, cur[1] + cj
+                if ni >= 0 and ni < nrows and nj >= 0 and nj < ncols \
```

Comment:
This long if condition should be a function which return boolean for readability